DIP/DIP/Models/CardPaymentService.cs

```diff
@@ -0,0 +1,22 @@
+﻿using DIP.Interfaces;
+
+namespace DIP
+{
+    public class CardPaymentService<CardPayment> : IPaymentService<CardPayment> where CardPayment : IPayment
+    {
+        public string Pay(CardPayment payment )
+        {
+            if (ValidDiscount(payment.Discount)) {
+                double totalAfterDiscount = payment.Total - (payment.Total * payment.Total / 100);
+                return $"Payment Successful. Total: {payment.Total}, After Discount {totalAfterDiscount}";
+            } else {
+                throw new Exception("Discount is not Valid");
+            }
+        }
+
+        public bool ValidDiscount(double discount)
+        {
+            return discount.IsBetween(0, 100);
```

Comment:
you don't have to do this, but I would rather if you use Guards (you car write your own, or use libraries).
it's easy to read, and contains almost all the straight forward cases, and you can see it's used a lot in many open source projects.
and usually it throws an exception directly without the need to check with if clause and throw the exception manually.